File: mobile/lib/models/journey.dart
class Journey {
  final String id;
  final String userId;
  final String status;
  final String? destinationName;
  final double startLat;
  final double startLon;
  final double? currentLat;
  final double? currentLon;
  final double? destinationLat;
  final double? destinationLon;
  final DateTime startTime;
  final DateTime? endTime;
  final double totalDistanceKm;
  final String maxReportedRisk;

  Journey({
    required this.id,
    required this.userId,
    required this.status,
    this.destinationName,
    required this.startLat,
    required this.startLon,
    this.currentLat,
    this.currentLon,
    this.destinationLat,
    this.destinationLon,
    required this.startTime,
    this.endTime,
    required this.totalDistanceKm,
    required this.maxReportedRisk,
  });

  factory Journey.fromJson(Map<String, dynamic> json) {
    return Journey(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      status: json['status'] as String,
      destinationName: json['destination_name'] as String?,
      startLat: (json['start_lat'] as num).toDouble(),
      startLon: (json['start_lon'] as num).toDouble(),
      currentLat: (json['current_lat'] as num?)?.toDouble(),
      currentLon: (json['current_lon'] as num?)?.toDouble(),
      destinationLat: (json['destination_lat'] as num?)?.toDouble(),
      destinationLon: (json['destination_lon'] as num?)?.toDouble(),
      startTime: DateTime.parse(json['start_time'] as String),
      endTime: json['end_time'] != null ? DateTime.parse(json['end_time'] as String) : null,
      totalDistanceKm: (json['total_distance_km'] as num?)?.toDouble() ?? 0.0,
      maxReportedRisk: json['max_reported_risk_encountered'] as String? ?? 'lower_reported_risk',
    );
  }
}

class JourneyShareInfo {
  final String id;
  final String journeyId;
  final String shareToken;
  final DateTime expiresAt;
  final bool isRevoked;
  final int viewsCount;
  final String? shareUrl;

  JourneyShareInfo({
    required this.id,
    required this.journeyId,
    required this.shareToken,
    required this.expiresAt,
    required this.isRevoked,
    required this.viewsCount,
    this.shareUrl,
  });

  factory JourneyShareInfo.fromJson(Map<String, dynamic> json) {
    return JourneyShareInfo(
      id: json['id'] as String,
      journeyId: json['journey_id'] as String,
      shareToken: json['share_token'] as String,
      expiresAt: DateTime.parse(json['expires_at'] as String),
      isRevoked: json['is_revoked'] as bool? ?? false,
      viewsCount: json['views_count'] as int? ?? 0,
      shareUrl: json['share_url'] as String?,
    );
  }
}

File: backend/app/schemas/journey.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from backend.app.models.journey import JourneyStatus


class JourneyStart(BaseModel):
    destination_name: Optional[str] = Field(None, max_length=255)
    start_lat: float = Field(..., ge=-90.0, le=90.0)
    start_lon: float = Field(..., ge=-180.0, le=180.0)
    destination_lat: Optional[float] = Field(None, ge=-90.0, le=90.0)
    destination_lon: Optional[float] = Field(None, ge=-180.0, le=180.0)


class JourneyLocationPing(BaseModel):
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    speed: Optional[float] = None
    accuracy: Optional[float] = None


class JourneyEnd(BaseModel):
    final_lat: Optional[float] = None
    final_lon: Optional[float] = None


class JourneyLocationOut(BaseModel):
    id: str
    latitude: float
    longitude: float
    speed: Optional[float]
    accuracy: Optional[float]
    timestamp: datetime

    class Config:
        from_attributes = True


class JourneyOut(BaseModel):
    id: str
    user_id: str
    status: JourneyStatus
    destination_name: Optional[str]
    start_lat: float
    start_lon: float
    current_lat: Optional[float]
    current_lon: Optional[float]
    destination_lat: Optional[float]
    destination_lon: Optional[float]
    start_time: datetime
    end_time: Optional[datetime]
    total_distance_km: float
    max_reported_risk_encountered: str
    created_at: datetime

    class Config:
        from_attributes = True


class JourneyShareCreate(BaseModel):
    duration_minutes: int = Field(30, ge=5, le=1440)  # 5 mins to 24 hours


class JourneyShareOut(BaseModel):
    id: str
    journey_id: str
    share_token: str
    expires_at: datetime
    is_revoked: bool
    views_count: int
    created_at: datetime
    share_url: Optional[str] = None

    class Config:
        from_attributes = True


class PublicShareAccessOut(BaseModel):
    journey_id: str
    status: JourneyStatus
    destination_name: Optional[str]
    current_lat: Optional[float]
    current_lon: Optional[float]
    start_time: datetime
    last_update_time: Optional[datetime]
    total_distance_km: float
    expires_at: datetime
    is_valid: bool
